# BlinkIT Grocery Data Analytics Project



## Data Cleaning & Profiling

### Objective
The goal of this phase is to understand the dataset structure and identify data quality issues before cleaning and analysis.

### Tasks
- Load dataset
- Understand dataset structure
- Identify missing values
- Identify duplicate records
- Detect inconsistent categories
- Detect business logic issues

In [14]:
import pandas as pd
import numpy as np

## Load Dataset

### Reading the intentionally messy dataset created for real-world practice.

In [15]:
df = pd.read_csv('../Dataset/blinkit_dataset.csv')
df.head()

,Item_Identifier,Item_Weight,Item_Fat_Content,Item_Visibility,Item_Type,Item_MRP,Outlet_Identifier,Outlet_Establishment_Year,Outlet_Size,Outlet_Location_Type,Outlet_Type,Item_Outlet_Sales
0,FDA15,9.30,LF,0.016047,Dairy,249.8092,OUT049,1999,Medium,Tier 1,Supermarket Type1,3735.1380
1,DRC01,5.92,reg,0.019278,Soft Drinks,48.2692,OUT018,2009,Medium,Tier 3,Supermarket Type2,443.4228
2,FDN15,17.50,LF,0.016760,Meat,141.6180,OUT049,1999,Medium,Tier 1,Supermarket Type1,2097.2700
3,FDX07,19.20,reg,0.000000,Fruits and Vegetables,182.0950,OUT010,1998,NaN,Tier 3,Grocery Store,732.3800
4,NCD19,8.93,LF,0.000000,Household,53.8614,OUT013,1987,High,Tier 3,Supermarket Type1,994.7052


## Dataset Overview

### Understanding the dimensions and structure of the dataset.

In [16]:
print("Rows and Columns:", df.shape)
df.info()

Rows and Columns: (8543, 12)
<class 'pandas.DataFrame'>
RangeIndex: 8543 entries, 0 to 8542
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8543 non-null   str    
 1   Item_Weight                7034 non-null   float64
 2   Item_Fat_Content           8543 non-null   str    
 3   Item_Visibility            8543 non-null   float64
 4   Item_Type                  8543 non-null   str    
 5   Item_MRP                   8493 non-null   float64
 6   Outlet_Identifier          8543 non-null   str    
 7   Outlet_Establishment_Year  8543 non-null   int64  
 8   Outlet_Size                6130 non-null   str    
 9   Outlet_Location_Type       8543 non-null   str    
 10  Outlet_Type                8543 non-null   str    
 11  Item_Outlet_Sales          8543 non-null   float64
dtypes: float64(4), int64(1), str(7)
memory usage: 801.0 KB


## Missing Values Analysis

### Identifying columns that contain missing values in the dataset.
### Missing values can affect analysis and machine learning results.

In [17]:
df.isnull().sum()

Item_Identifier                 0
Item_Weight                  1509
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                       50
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2413
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

## Missing Values Findings

### Observations

- Item_Weight contains 1509 missing values.
- Item_MRP contains 50 missing values.
- Outlet_Size contains 2413 missing values.

### Initial Decision

- Numerical columns may use median-based filling.
- Categorical columns may use mode or business-based filling.
- Rows will not be dropped initially because of large data loss risk.

## Duplicate Records Analysis

### Identifying duplicate records in the dataset.

Duplicate values can create incorrect calculations and misleading business insights.

In [18]:
df.duplicated().sum()

np.int64(19)

## Unique Values Analysis

### Understanding unique categories and identifying inconsistent values.

### This helps detect:
- Wrong spellings
- Inconsistent categories
- Unexpected values
- Data quality issues

In [19]:
for col in df.columns:
    print(col)
    print(df[col].unique())
    print("-" * 50)

Item_Identifier
<StringArray>
['FDA15', 'DRC01', 'FDN15', 'FDX07', 'NCD19', 'FDP36', 'FDO10', 'FDP10',
 'FDH17', 'FDU28',
 ...
 'FDB46', 'NCX17', 'FDH31', 'FDX13', 'NCU53', 'FDD28', 'FDU43', 'NCF55',
 'NCW30', 'NCW05']
Length: 1559, dtype: str
--------------------------------------------------
Item_Weight
[ 9.3    5.92  17.5   19.2    8.93  10.395 13.65     nan 16.2   11.8
 18.5   15.1   17.6   16.35   9.    13.35  18.85  14.6   13.85  13.
  7.645 11.65   5.925 19.25  18.6   18.7   17.85  10.     8.85   9.8
 13.6   21.35  12.15   6.42  19.6   15.85   7.39  10.195  9.895 10.895
  7.905  9.195  8.365  7.97  17.7   19.35   8.645 15.6   18.25   7.855
  7.825  8.39  12.85  19.     5.905  7.76  15.5    6.055  6.305 20.85
 20.75   8.895 19.7    8.75  13.3    8.31  16.75  19.75  17.1   10.5
  6.635 14.15   8.89   9.1    7.5   16.85   7.485 11.6   12.65  20.25
  8.6   12.6    8.88  20.5   13.5    7.235  6.92   8.02  12.8   16.6
 14.    16.    21.25   7.365 18.35   5.465  7.27   6.155 19.5   15.

In [20]:
df['Item_Fat_Content'].value_counts()

Item_Fat_Content
LF         5418
reg        3013
low fat     112
Name: count, dtype: int64

In [21]:
df['Outlet_Size'].value_counts()

Outlet_Size
Medium    2802
Small     2394
High       934
Name: count, dtype: int64

In [22]:
df['Outlet_Location_Type'].value_counts()

Outlet_Location_Type
Tier 3    3359
Tier 2    2790
Tier 1    2394
Name: count, dtype: int64

## Category Standardization
### Cleaning inconsistent categorical values to maintain data quality and ensure accurate analysis.

In [23]:
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace({
    'LF':"Low Fat",
    "low fat":"Low Fat",
    "reg":"Regular"
})
df['Item_Fat_Content'].value_counts()

Item_Fat_Content
Low Fat    5530
Regular    3013
Name: count, dtype: int64

## Duplicate Record Cleaning

### Removing duplicate records to prevent double-counting and incorrect business calculations.

In [24]:
df = df.drop_duplicates()
df.duplicated().sum()

np.int64(0)

## Missing Value Treatment

### Handling missing values in numerical and categorical columns to maintain data quality.

In [25]:
df[['Item_Weight','Item_MRP']].median()

Item_Weight     12.6000
Item_MRP       142.9312
dtype: float64

In [26]:
df['Item_Weight'] = df['Item_Weight'].fillna(df['Item_Weight'].median())

In [27]:
df['Item_MRP'] = df['Item_MRP'].fillna(df['Item_MRP'].median())

In [28]:
df[['Item_Weight','Item_MRP']].isnull().sum()

Item_Weight    0
Item_MRP       0
dtype: int64

## Categorical Missing Value Treatment

### Filling missing values in categorical columns using the most frequent category (mode).

In [29]:
df['Outlet_Size'].isnull().sum()

np.int64(2410)

In [30]:
df['Outlet_Size'].mode()

0    Medium
Name: Outlet_Size, dtype: str

In [31]:
df['Outlet_Size'] = df['Outlet_Size'].fillna(
    df['Outlet_Size'].mode()[0]
)
df['Outlet_Size'].isnull().sum()

np.int64(0)

In [32]:
df['Outlet_Size'].unique()

<StringArray>
['Medium', 'High', 'Small']
Length: 3, dtype: str

In [33]:
df['Outlet_Size'].value_counts()

Outlet_Size
Medium    5203
Small     2388
High       933
Name: count, dtype: int64

## Data Validation

### Verifying whether all missing values have been handled successfully.

In [34]:
df.isnull().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64